File: workmate-server/src/utils/whitelist-matcher.ts
import type { RuleType } from '../types'

export type { RuleType }

export interface WhitelistRule {
  ruleType: RuleType
  ruleValue: string
}

export interface MatcherResult {
  matched: boolean
  matchedRule?: WhitelistRule
}

export function matchWhitelist(jobId: string, rules: WhitelistRule[]): MatcherResult {
  for (const rule of rules) {
    switch (rule.ruleType) {
      case 'list': {
        const ids = rule.ruleValue.split(',').map((id) => id.trim())
        if (ids.includes(jobId)) {
          return { matched: true, matchedRule: rule }
        }
        break
      }
      case 'range': {
        const parts = rule.ruleValue.split('-')
        if (parts.length === 2) {
          const start = parts[0].trim()
          const end = parts[1].trim()
          if (jobId >= start && jobId <= end) {
            return { matched: true, matchedRule: rule }
          }
        }
        break
      }
      case 'prefix': {
        const prefix = rule.ruleValue.replace(/\*$/, '')
        if (jobId.startsWith(prefix)) {
          return { matched: true, matchedRule: rule }
        }
        break
      }
      case 'suffix': {
        const suffix = rule.ruleValue.replace(/^\*/, '')
        if (jobId.endsWith(suffix)) {
          return { matched: true, matchedRule: rule }
        }
        break
      }
    }
  }
  return { matched: false }
}

export function matchAnyRule(jobId: string, rules: WhitelistRule[]): boolean {
  return matchWhitelist(jobId, rules).matched
}

File: workmate-server-java/src/main/java/com/workmate/server/util/WhitelistMatcher.java
package com.workmate.server.util;

import java.util.List;

public class WhitelistMatcher {

    public static boolean matchAnyRule(String jobId, List<WhitelistRule> rules) {
        return rules.stream().anyMatch(rule -> matchRule(jobId, rule));
    }

    private static boolean matchRule(String jobId, WhitelistRule rule) {
        return switch (rule.getRuleType()) {
            case "list" -> {
                String[] ids = rule.getRuleValue().split(",");
                for (String id : ids) {
                    if (id.trim().equals(jobId)) {
                        yield true;
                    }
                }
                yield false;
            }
            case "range" -> {
                String[] parts = rule.getRuleValue().split("-");
                if (parts.length == 2) {
                    String start = parts[0].trim();
                    String end = parts[1].trim();
                    yield jobId.compareTo(start) >= 0 && jobId.compareTo(end) <= 0;
                }
                yield false;
            }
            case "prefix" -> {
                String prefix = rule.getRuleValue().replaceAll("\\*$", "");
                yield jobId.startsWith(prefix);
            }
            case "suffix" -> {
                String suffix = rule.getRuleValue().replaceAll("^\\*", "");
                yield jobId.endsWith(suffix);
            }
            default -> false;
        };
    }

    public record WhitelistRule(String ruleType, String ruleValue) {
        public String getRuleType() { return ruleType; }
        public String getRuleValue() { return ruleValue; }
    }
}
